File: sol/530.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def getAdj(self, root: Optional[TreeNode]):
        assert root is not None
        minVal = root.val
        maxVal = root.val
        diff = 100001
        if root.left is not None:
            subTuple = self.getAdj(root.left)
            minVal = subTuple[0]
            diff = min(diff, root.val - subTuple[1], subTuple[2])
        if root.right is not None:
            subTuple = self.getAdj(root.right)
            maxVal = subTuple[1]
            diff = min(diff, subTuple[0] - root.val, subTuple[2])
        return (minVal, maxVal, diff)


    def getMinimumDifference(self, root: Optional[TreeNode]) -> int:
        return self.getAdj(root)[2]

File: sol/530/530.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
struct Solution {
    int getSub(TreeNode *root, int &minVal, int &maxVal){
        assert(root != nullptr);
        int diff = 100001, subMin, subMax;
        minVal = root->val;
        maxVal = root->val;
        if (root->left){
            diff = min(diff, this->getSub(root->left, minVal, subMax));
            diff = min(diff, root->val - subMax);
        }
        if (root->right){
            diff = min(diff, this->getSub(root->right, subMin, maxVal));
            diff = min(diff, subMin - root->val);
        }
        return diff;
    }

    int getMinimumDifference(TreeNode* root) {
        int minVal, maxVal;
        return this->getSub(root, minVal, maxVal);
    }
};
